src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp

```diff
@@ -1202,3 +1205,129 @@ void CascadiaSettings::_resolveDefaultProfile() const
     // Use the first profile as the new default.
     GlobalSettings().DefaultProfile(_allProfiles.GetAt(0).Guid());
 }
+
+// Method Description:
+// - Iterates through the "newTabMenu" entries and for ProfileEntries resolves the "profile"
+//   fields, which can be a profile name, to a GUID and stores it back.
+// - It finds any "source" entries and finds all profiles generated by that source
+// - Lastly, it finds any "remainingProfiles" entries and stores which profiles they
+//   represent (those that were not resolved before). It adds a warning when
+//   multiple of these entries are found.
+void CascadiaSettings::_resolveNewTabMenuProfiles() const
+{
+    Model::RemainingProfilesEntry remainingProfilesEntry = nullptr;
+
+    // The TerminalPage needs to know which profile has which profile ID. To prevent
+    // continuous lookups in the _activeProfiles vector, we create a map <int, Profile>
+    // to store these indices in-flight.
+    auto remainingProfilesMap = std::map<int, Model::Profile>{};
+    auto activeProfileCount = gsl::narrow_cast<int>(_activeProfiles.Size());
+    for (auto profileIndex = 0; profileIndex < activeProfileCount; profileIndex++)
+    {
+        remainingProfilesMap.emplace(profileIndex, _activeProfiles.GetAt(profileIndex));
+    }
+
+    // We keep track of the "remaining profiles" - those that have not yet been resolved
+    // in either a "profile" or "source" entry. They will possibly be assigned to a
+    // "remainingProfiles" entry
+    auto remainingProfiles = single_threaded_map(std::move(remainingProfilesMap));
+
+    // We call a recursive helper function to process the entries
+    auto entries = _globals->NewTabMenu();
+    _resolveNewTabMenuProfilesSet(entries, remainingProfiles, remainingProfilesEntry);
+
+    // If a "remainingProfiles" entry has been found, assign to it the remaining profiles
+    if (remainingProfilesEntry != nullptr)
+    {
+        remainingProfilesEntry.Profiles(remainingProfiles);
+    }
+
+    // If the configuration does not have a "newTabMenu" field, GlobalAppSettings
+    // will return a default value containing just a "remainingProfiles" entry. However,
+    // this value is regenerated on every "get" operation, so the effect of setting
+    // the remaining profiles above will be undone. So only in the case that no custom
+    // value is present in GlobalAppSettings, we will store the modified default value.
+    if (!_globals->HasNewTabMenu())
+    {
+        _globals->NewTabMenu(entries);
+    }

```

Comment:
Okay so I had

```json
    "newTabMenu": [
        { "type": "remainingProfiles" },
        {
            "type": "folder",
            "name": "misc",
            "entries": [
                { "type": "profile", "profile": "Ubuntu-18.04" },
                { "type": "profile", "profile": "Ubuntu-22.04" },
                { "type": "profile", "profile": "AlmaLinux-8" }
            ]
        }
    ],
```

and it roundtripped as 

```json
    "newTabMenu": 
    [
        {
            "type": "remainingProfiles"
        },
        {
            "entries": 
            [
                {
                    "profile": "{c6eaf9f4-32a7-5fdc-b5cf-066e8a4b1e40}",
                    "type": "profile"
                },
                {
                    "profile": "{17bf3de4-5353-5709-bcf9-835bd952a95e}",
                    "type": "profile"
                },
                {
                    "profile": "{809c0a51-3bb6-5c2b-93d9-741099225673}",
                    "type": "profile"
                }
            ],
            "icon": null,
            "name": "misc",
            "type": "folder"
        }
    ],
```

And that's totally fine by me.

Reply:
Yes this part is just about the remainingprofiles part.

What was happening: in _globals, a default value for the NewTabMenu is returned: a simple array with just a single RemainingProfiles entry. However, since no value was present, every get operation on this field in _globals would recreate the object instead of returning a cached default. Hence; when processing the menu and computing which profiles need to be rendered, this information is stored in the TabMenu structure which is then recreated when we are planning to actually use it.

So _just_ in the case that no custom menu is set, we temporarily set a "custom" value which is just the default version but modified for runtime results.

I hope this clears things up :P